crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -67,6 +72,40 @@ pub type DatabaseProviderRO<DB> = DatabaseProvider<<DB as Database>::TX>;
 #[derive(Debug)]
 pub struct DatabaseProviderRW<DB: Database>(pub DatabaseProvider<<DB as Database>::TXMut>);
 
+impl<DB> DatabaseProviderRW<DB>
+where
+    DB: Database,
+{
+    /// Returns a provider with the given `DbTxMut` inside, which allows fetching and updating
+    /// data from the database using different types of providers.
+    #[track_caller]
+    pub fn with_tx(
+        tx: <DB as Database>::TXMut,
+        chain_spec: Arc<ChainSpec>,
+        snapshot_provider: &Option<Arc<SnapshotProvider>>,
+    ) -> ProviderResult<DatabaseProviderRW<DB>> {
+        let mut provider = DatabaseProvider::new_rw(tx, chain_spec.clone());
+
+        if let Some(snapshot_provider) = snapshot_provider {
+            provider = provider.with_snapshot_provider(snapshot_provider.clone());
+        }
```

Comment:
there should also be a function on the db provider that accepts all args

Reply:
just added one, this method now calls that method